Add TFM check for `Publish-Trimmed-ReadyToRun-SingleFile`

swaroop-sridhar opened the pull request:
Trying to publish a `.net core 2.1` app as a single file fails with the `PlaceHolderNotFoundInAppHostException` generated by the `HostWriter`.

Publishing apps as a single file is only supported in `.net core 3` and higher.
So, this change adds a version check in the SDK.

Similarly, the check is also added for the `PublishReadyToRun` and `PublishTrimmed` options.

Fixes https://github.com/dotnet/sdk/issues/3728

mmitche committed:
[release/3.0.1xx] Rebrand for 3.0.101 (#3681)

dotnet-maestro[bot] committed:
[release/3.0.1xx] Update dependencies from dotnet/arcade (#3674)

* Update dependencies from https://github.com/dotnet/arcade build 20190924.3

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19474.3
- Microsoft.DotNet.SignTool - 1.0.0-beta.19474.3

* Add feed for dotnet-tools and remove unneeded package sources

nguerrera committed:
Merge branch 'release/3.1.1xx' into merges/release/3.0.1xx-to-release/3.1.1xx

nguerrera committed:
Flow available RIDs to R2R via targeting pack, not runtime pack (#3686)

nguerrera committed:
Revert "Remove perf from builds (#3654)" (#3676)

This reverts commit b3a343bf8ed65f5208c5bb86c632856c5d107b1d.

nguerrera committed:
Merge pull request #3687 from dotnet/merges/release/3.0.1xx-to-release/3.1.1xx

Merge release/3.0.1xx to release/3.1.1xx

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3594)

* Update dependencies from https://github.com/dotnet/core-setup build 20190927.8

- Microsoft.NET.HostModel - 3.1.0-preview1.19477.8

Commit:
Better error message when c++ targeting TFM lower than 3.1 (#3689)

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3692)

* Update dependencies from https://github.com/dotnet/core-setup build 20190927.22

- Microsoft.NET.HostModel - 3.1.0-preview1.19477.22

* Update dependencies from https://github.com/dotnet/core-setup build 20190930.12

- Microsoft.NET.HostModel - 3.1.0-preview1.19480.12

* Update dependencies from https://github.com/dotnet/core-setup build 20190930.22

- Microsoft.NET.HostModel - 3.1.0-preview1.19480.22

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3699)

* Update dependencies from https://github.com/dotnet/core-setup build 20190930.25

- Microsoft.NET.HostModel - 3.1.0-preview1.19480.25

* Update dependencies from https://github.com/dotnet/core-setup build 20190930.27

- Microsoft.NET.HostModel - 3.1.0-preview1.19480.27

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191006.06

- Microsoft.NET.HostModel - 3.0.1

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3702)

* Update dependencies from https://github.com/dotnet/core-setup build 20191001.9

- Microsoft.NET.HostModel - 3.1.0-preview1.19501.9

* Update dependencies from https://github.com/dotnet/core-setup build 20191001.13

- Microsoft.NET.HostModel - 3.1.0-preview1.19501.13

* Update dependencies from https://github.com/dotnet/core-setup build 20191001.15

- Microsoft.NET.HostModel - 3.1.0-preview1.19501.15

* Update dependencies from https://github.com/dotnet/core-setup build 20191001.22

- Microsoft.NET.HostModel - 3.1.0-preview1.19501.22

* Update dependencies from https://github.com/dotnet/core-setup build 20191001.24

- Microsoft.NET.HostModel - 3.1.0-preview1.19501.24

* Move to open pool

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191003.14 (#3711)

- Microsoft.NET.HostModel - 3.1.0-preview1.19503.14

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191004.2 (#3712)

- Microsoft.NET.HostModel - 3.1.0-preview1.19504.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191004.13 (#3715)

- Microsoft.NET.HostModel - 3.1.0-preview1.19504.13

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191004.15 (#3716)

- Microsoft.NET.HostModel - 3.1.0-preview1.19504.15

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3717)

* Update dependencies from https://github.com/dotnet/core-setup build 20191004.18

- Microsoft.NET.HostModel - 3.1.0-preview1.19504.18

* Update dem deps

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191005.4 (#3720)

- Microsoft.NET.HostModel - 3.1.0-preview1.19505.4

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191005.5 (#3721)

- Microsoft.NET.HostModel - 3.1.0-preview1.19505.5

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191006.1 (#3722)

- Microsoft.NET.HostModel - 3.1.0-preview1.19506.1

nguerrera committed:
Update stage 0

nguerrera committed:
Add test feed

nguerrera committed:
Download runtime packs for latest runtime version by default

nguerrera committed:
Change linux machine pool and remove Ubuntu 14.04

dsplaisted committed:
Merge pull request #3703 from dotnet/darc-release/3.0.1xx-0cfb7de9-3d74-408e-962e-c4c3b01c360d

[release/3.0.1xx] Update dependencies from dotnet/core-setup

nguerrera committed:
Merge branch 'release/3.0.1xx' into merge-3.0-to-3.1

nguerrera committed:
Merge pull request #3733 from nguerrera/merge-3.0-to-3.1

Merge 3.0.1xx to 3.1.1xx

vitek-karas committed:
Implement writing includedFrameworks references for self-contained apps (#3697)

For self-contained apps which are of the right version (to maintain backward compat), 
write the framework references to a new includedFrameworks property in runtime 
config.

dsplaisted committed:
Update branding to 3.1.100-preview2

dsplaisted committed:
Merge pull request #3747 from dsplaisted/update-branding

Update branding to 3.1.100-preview2

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3740)

* Update dependencies from https://github.com/dotnet/core-setup build 20191010.20

- Microsoft.NET.HostModel - 3.1.0-preview2.19510.20

* Update dependencies from https://github.com/dotnet/core-setup build 20191010.21

- Microsoft.NET.HostModel - 3.1.0-preview2.19510.21

* Update dependencies from https://github.com/dotnet/core-setup build 20191011.6

- Microsoft.NET.HostModel - 3.1.0-preview2.19511.6

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191015.16 (#3756)

- Microsoft.NET.HostModel - 3.1.0-preview2.19515.16

swaroop-sridhar committed:
Add TFM check for `Publish-Trimmed-ReadyToRun-SingleFile`

The options `PublishSingleFile`, `PublishReadyToRun`, and `PublishTrimmed` are only supported when targetting netcoreapp3.0 or later.

Trying to publish to a different target (say `.netcoreapp2.1`) today:
* `PublishSingleFile` fails with the `PlaceHolderNotFoundInAppHostException` generated by the `HostWriter`.
* `PublishReadyToRun` and `PublishTrimmed` silently fail to turn on, but the `publish` itself suceeds.

This change adds an explicit TFM check to generate error/warnings for non-conforming targets.

`PublishSingleFile`, `PublishReadyToRun`, and `PublishTrimmed` require the following conditions to be true:
* `TargetFramework` is `netcoreapp`
* `TargetFrameworkVersion` is at least `3.0`
* The app is an executable (`OutputType=exe`)

If any of the conditions fail:
* `PublishSingleFile` fails with an appropriate error
* `PublishReadyToRun` and `PublishTrimmed` issue warnings, but publish itself succeeds.
  This is because:
    * The restriction against `classlibs` / `netstandard` frameworks may be relaxed in future releases
    * The version check should ideally be a failure, but they are warnings for maximum compatibility with `3.0` release.

Fixes https://github.com/dotnet/sdk/issues/3728

swaroop-sridhar commented:
CC: @nguerrera  please review.

swaroop-sridhar commented:
CC: @fadimounir @sbomer  @vitek-karas

fadimounir commented:
LGTM

fadimounir commented:
BTW, we shouldn't be executing R2R for any version less than 3.0 today.
see: https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L214

swaroop-sridhar commented:
@fadimounir Shouldn't we check for these conditions, and fail ... rather than silently not generate R2R code?

fadimounir commented:
Yes it's better to either fail, or emit a warning that it's not supported and skip the R2R/Trimming/SingleExe step. I'm good with either options

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:
```diff
@@ -142,6 +142,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+
+    <NETSdkError Condition="'$(PublishTrimmed)' == 'true' and '$(_TargetFrameworkVersionWithoutV)' &lt; '3.0'"
+             ResourceName="PublishTrimmedRequiresVersion30" />
+
```
These should also check TargetFrameworkIdentifier is .NETCoreApp.

nguerrera commented:
What release do you want this in? Should it sit until 5 or go out with 3.1 soon? I think 3.1. Can you rebase and retarget to release/3.1.1xx? We don't dual checkin master and 3.1, we rely on merge from 3.1 to master.

swaroop-sridhar commented:
OK Thanks @nguerrera, I'll put the PR for 3.1xx branch

nguerrera commented:
For R2R and ILLink, this is a breaking change. They both no-op now on less than 3.0.

I think this is better, but it makes me re-question a bit if it should go in 3.1.

@livarcocc @dsplaisted What do you think?

One minor scenario would be if you multi-target and want to use these with your 3.0+ targets, but not the others. You would now have to add a condition on TargetFramework. This is arguably more obvious to reader of project file what would happen, but it is breaking.

swaroop-sridhar commented:
@nguerrera, how about making it a warning for 3.0 for R2R and Trimming in 3.1? 

Also, how about adding warnings (for R2R and Trimming) if the following checks fail? 
`'$(TargetFrameworkIdentifier)' == '.NETCoreApp'` ` '$(_IsExecutable)' == 'true'"` 
That is, emitting a warning, rather than silently failing to comply with the option? 

For R2R and IL Trimming, I think the above checks should only generate warnings, because they may be accomodated in future versions (ex: R2R code for libs).

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20191016.1 (#3757)

- Microsoft.NET.HostModel - 3.1.0-preview2.19516.1

dotnet-maestro[bot] committed:
[release/3.1.1xx] Update dependencies from dotnet/core-setup (#3761)

* Update dependencies from https://github.com/dotnet/core-setup build 20191016.15

- Microsoft.NET.HostModel - 3.1.0-preview2.19516.15

* Update dependencies from https://github.com/dotnet/core-setup build 20191016.18

- Microsoft.NET.HostModel - 3.1.0-preview2.19516.18

* Update dependencies from https://github.com/dotnet/core-setup build 20191017.4

- Microsoft.NET.HostModel - 3.1.0-preview2.19517.4

nguerrera commented:
Replaced by #3766
